### ISSUE TITLE
report unsafe operator type on subclasses

### ISSUE THREAD

**Issue by KotlinIsland** [astral-sh/ty#630]
### Summary

```py
class A:
    def __add__(self, other: A) -> int:
        return 1

class B(A):
    def __radd__(self, other: A) -> str:
        return "a"

def f(a1: A, a2: A):
    x: int = a1 + a2
    print(x)

f(A(), B())
```

here, the right hand subtype rule will be used, and `B.__radd__` will be called first, resulting in `"a"`, not `1`

if the supertype additionally defines a `__radd__`, then it can be incompatible with it's `__add__`, as all the information about the operation will be known

to ensure safety we must report when the subtypes `__radd__` in incompatable with the supertypes `__add__`, when the super type does not define a `__radd__`

this also applies to inplace operators:
```py
class A:
    def __add__(self, other: A) -> int:
        return 1
class B(A):
    def __iadd__(self, other: A) -> str:   # expect error: `__iadd__` must be compatable with `A.__add__`, as `A` does not define `__iadd__`
        return "a"

def f(a1: A, a2: A):
    a = a1
    a += a2
    print(a)  # static: int, runtime: "a"
f(B(), A())
```
note that the righthand subtype rule applies __after__ `__iadd__`

**Comment by carljm** [comment]
For reference, mypy and pyright both also give `int` here; mypy also gives the "unsafe operator definitions" diagnostic requested in this issue.

**Comment by KotlinIsland** [comment]
mypy doesn't fully meet the request here:
```py
class A:
    def __add__(self, other: A) -> int:
        return 1
    def __radd__(self, other: A) -> str:
        return "a"
class B(A):
    def __radd__(self, other: A) -> str:  # error: Signatures of "__radd__" of "B" and "__add__" of "A" are unsafely overlapping
        return "a"
```

this can be perfectly valid, if the analysis of the operation is modified to account for it
